sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -83,20 +83,35 @@ class EquivalentExpressions {
    * Adds only expressions which are common in each of given expressions, in a recursive way.
    * For example, given two expressions `(a + (b + (c + 1)))` and `(d + (e + (c + 1)))`,
    * the common expression `(c + 1)` will be added into `equivalenceMap`.
+   *
+   * Note that as we don't know in advance if any child node of an expression will be common
+   * across all given expressions, we count all child nodes when looking through the given
+   * expressions. But when we call `addExprTree` to add common expressions into the map, we
+   * will add recursively the child nodes. So we need to filter the child expressions first.
+   * For example, if `((a + b) + c)` and `(a + b)` are common expressions, we only add
+   * `((a + b) + c)`.
    */
   private def addCommonExprs(
       exprs: Seq[Expression],
       addFunc: Expression => Boolean = addExpr): Unit = {
     val exprSetForAll = mutable.Set[Expr]()
```

Comment:
One potentially unrelated thing I just noticed, do we need to keep track of all of the `Expressions` here as well (as in an `Expr -> Seq[Expression]` map)? It's really basically keeping the first `Expression` found, but the codegen looks like it uses the `Expression` hash (versus the semantic hash) to lookup subexpressions.

Reply:
You mean `equivalenceMap`?

Reply:
I don't mean add it directly to that here. I'm just thinking of a really stupid example, `when((col + 1) > 0, col + 1).otherwise(1 + col)`. Wouldn't `col + 1` and `1 + col` resolve as a common expression because they're semantically equal, but only `col + 1` is added to `equivalenceMap`, so during codegen `1 + col` wouldn't be resolved to the subexpression?

Reply:
`col + 1` and `1 + col` will both be recognized as subexpression.

Reply:
Yeah but won't the codgen stage not replace `1 + col` since only `col + 1` will be added to the `equivalenceMap` entry for `Expr(col + 1)`? For non commonExprs cases, both would be in `equivalenceMap` so that the codegen stage maps both of those expressions to the resulting subexpression. Again, not super related to this PR, but was the easiest place to ask

Reply:
Both `1 + col` and `col + 1` will be replaced with the extracted subexpression during codege. We don't just look of key at `equivalenceMap` when replacing with subexpression.